src/provider/provider_tracking.c

```diff
@@ -85,20 +231,50 @@ static umf_result_t umfMemoryTrackerRemove(umf_memory_tracker_handle_t hTracker,
     // Every umfMemoryTrackerAdd(..., ptr, ...) should have a corresponding
     // umfMemoryTrackerRemove call with the same ptr value.
 
-    void *value = critnib_remove(hTracker->alloc_segments_map, (uintptr_t)ptr);
+    umf_result_t umf_result = UMF_RESULT_ERROR_UNKNOWN;
+    tracker_alloc_info_t *parent_value = NULL;
+    uintptr_t parent_key = 0;
+    int level = 0;
+
+    int ret = utils_mutex_lock(&hTracker->mutex);
+    if (ret) {
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    // Find the most nested (on the highest level) entry in the map
+    // with the `ptr` key and with no children - only such entry can be removed.
+    tracker_alloc_info_t *value = get_most_nested_alloc_segment(
+        hTracker, ptr, &level, &parent_key, &parent_value, 1 /* no_children */);
     if (!value) {
         LOG_ERR("pointer %p not found in the alloc_segments_map", ptr);
-        return UMF_RESULT_ERROR_UNKNOWN;
+        goto err_unlock;
     }
 
-    tracker_alloc_info_t *v = value;
+    assert(level < MAX_LEVELS_OF_ALLOC_SEGMENT_MAP);
+    value = critnib_remove(hTracker->alloc_segments_map[level], (uintptr_t)ptr);
+    assert(value);
 
-    LOG_DEBUG("memory region removed: tracker=%p, ptr=%p, size=%zu",
-              (void *)hTracker, ptr, v->size);
+    LOG_DEBUG("memory region removed: tracker=%p, level=%i, pool=%p, ptr=%p, "
+              "size=%zu",
+              (void *)hTracker, level, value->pool, ptr, value->size);
+
+    if (parent_value) {
+        LOG_DEBUG(
+            "child #%zu removed from memory region: tracker=%p, level=%i, "
+            "pool=%p, ptr=%p, size=%zu",
+            parent_value->n_children, (void *)hTracker, level - 1,
+            (void *)parent_value->pool, (void *)parent_key, parent_value->size);
+        parent_value->n_children--;
```

Comment:
What if parent_value->n_children becomes 0? Don't we have to remove the parent as well recursively?

Reply:
We don't, only then it **can** be removed.